design/external-names.md

```diff
@@ -0,0 +1,47 @@
+# External Names Service
+
+_Status_: Draft
+
+This document looks to allow Contour to process Kubernetes service types `ExternalName` so that traffic can be proxied to dns hosts rather than only pod endpoints.
+
+## Goals
+
+- Allow for Contour to proxy traffic via Envoy to external dns endpoints
+- Utilize built-in Kubernetes object types (i.e. ExternalName service)
+
+## Non-goals
+
+- `externalIPs` will not be used, users who want to proxy to custom IP's should build their own service/endpoint object
+- Utilize a new CRD or ConfigMap style mapping
+
+## Background
+
+Contour currently watches Kubernetes services and endpoints, streaming them to via Envoy xDS CDS & EDS endpoints.
+Requests to envoy are routed directly to Kubernetes pods (i.e. endpoints).
+This proposal looks to implement Kubernetes service type `ExternalName` which allows for an `externalName` which is represented as a valid RFC-1123 hostname.
+
+## High-level Design
+
+Contour will need to watch for and process Kubernetes service types of `ExternalName` (https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.13/#servicespec-v1-core).
+When configuring EDS endpoints, for this service type Contour will pass the contents of the `externalName` field on the Kubernetes service to EDS instead of a PodIP.
+
+Additionally, users will need to configure ports on the service so Contour will understand how to proxy requests to the external endpoint.
+
+## Detailed Design
+
+- The `TCPService` struct in `internal/dag/dag.go` will have a new `externalName string` field added.
+- When services are added in `internal/dag/builder.go`, if `svc.Spec.Type == v1.ServiceTypeExternalName` then the value of `svc.Spec.ExternalName` will be set to the TCPService struct's `externalName`.
+- Since the service does not have any Kubernetes Endpoints, we cannot rely on having an EDS cluster to reference from the CDS cluster.
+For this type of service (i.e. `ExternalName`) when creating the cluster in `internal/envoy/cluster.go`, instead of setting a `EdsClusterConfig`, we'll configure a `LoadAssignment *ClusterLoadAssignment` which should allow us to specify the endpoint defined in `externalName` dynamically.
+
+## Alternatives
+
+Potentially, we could avoid using a Service of type `ExternalName` by simply just extending the `IngressRoute` CRD as proposed here: https://github.com/heptio/contour/issues/825#issue-386869426
+
+The downside to this approach is it couldn't be used with Ingress resources.
+
+## Security Considerations
```

Comment:
It might also be worth noting that users should probably pair this with the new upstream TLS validation feature (added in https://github.com/heptio/contour/pull/1045).

In a large percentage of deployments, this externalName upstream connection is more likely to be transiting untrusted networks.

Reply:
Yeah, this will give users the opportunity to have TLS between clusters.